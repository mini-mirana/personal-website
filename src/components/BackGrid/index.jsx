import { useEffect } from 'react'
import { useThree } from '@react-three/fiber'
// import { Plane } from '@react-three/drei'
import { Float } from '@react-three/drei'
import * as THREE from 'three'

export function BackGrid({ ...props }) {
  const { scene } = useThree()
  useEffect(() => {
    scene.fog = new THREE.FogExp2(0, 0.05)
  }, [scene])

  return (
    // <Plane args={[80, 80, 128, 128]} {...props}>
    //   <meshStandardMaterial color='#ea5455' wireframe side={THREE.DoubleSide} />
    // </Plane>
    <Float scale={0.75} rotationIntensity={0.01} floatIntensity={1} speed={3}>
      <mesh {...props}>
        <cylinderGeometry args={[5, 5, 500, 54, 64, true, 0, 6.283185307179586]} />
        <meshStandardMaterial color='#2ec4b6' wireframe side={THREE.DoubleSide} />
      </mesh>
    </Float>
  )
}

export default BackGrid
