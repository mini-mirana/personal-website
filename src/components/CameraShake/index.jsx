import React from 'react'
import { useFrame, useThree } from '@react-three/fiber'
import { SimplexNoise } from 'three-stdlib'

export function CameraShake({
  intensity = 1,
  decay,
  decayRate = 0.65,
  maxYaw = 0.1,
  maxPitch = 0.1,
  yawFrequency = 0.1,
  pitchFrequency = 0.1
}) {
  const camera = useThree((state) => state.camera)
  const intensityRef = React.useRef(intensity)
  const initialRotation = React.useRef(camera.rotation.clone())

  const [yawNoise] = React.useState(() => new SimplexNoise())
  const [pitchNoise] = React.useState(() => new SimplexNoise())

  const constrainIntensity = () => {
    if (intensityRef.current < 0 || intensityRef.current > 1) {
      intensityRef.current = intensityRef.current < 0 ? 0 : 1
    }
  }

  useFrame(({ clock }, delta) => {
    const shake = intensityRef.current ** 2
    const yaw = maxYaw * shake * yawNoise.noise(clock.elapsedTime * yawFrequency, 1)
    const pitch = maxPitch * shake * pitchNoise.noise(clock.elapsedTime * pitchFrequency, 1)

    camera.rotation.set(initialRotation.current.x + pitch, initialRotation.current.y + yaw, camera.rotation.z)

    if (decay && intensityRef.current > 0) {
      intensityRef.current -= decayRate * delta
      constrainIntensity()
    }
  })

  return null
}

export default CameraShake
