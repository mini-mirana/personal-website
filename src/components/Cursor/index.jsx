import * as THREE from 'three'
import React, { useEffect, useRef, useState } from 'react'
import { useFrame } from '@react-three/fiber'
import UseAnimations from 'react-useanimations'
import { openInNewTab } from '../../utils/utils'

export function Cursor({
  children,
  speed = 0.1,
  eps = 0.001,
  /* col = new THREE.Color(), */ vec = new THREE.Vector2(),
  icons,
  sections,
  p = new THREE.Vector3(),
  dom,
  frustum = new THREE.Frustum(),
  cameraViewProjectionMatrix = new THREE.Matrix4(),
  bbox = new THREE.Box3()
}) {
  const outer = useRef()
  const [hovered, hover] = useState()
  const [pos] = useState(() => new THREE.Vector2())
  const ref = useRef()
  const clicked = useRef()
  const [currentSection, setCurrentSection] = useState(sections[0]?.section)
  const clickSound = new Audio('/sound/type.mp3')
  const color = hovered ? 'rgb(100 116 139)' : '#fff'

  useEffect(() => {
    // Your code here
    window.addEventListener(
      'mousemove',
      (e) => {
        vec.set(e.pageX, e.pageY)
        if (Math.abs(pos.x - vec.x) > eps || Math.abs(pos.y - vec.y) > eps) {
          pos.lerp(vec, speed)
          outer.current.style.transform = `translate3d(${pos.x - 55}px,${pos.y - 55}px,0)`
        }
      },
      false
    )
  }, [])

  // useEffect(() => {
  //   clicked.current = ref.current.getObjectByName(params?.id)
  //   if (clicked.current) {
  //     clicked.current.parent.parent.updateWorldMatrix(true, true)
  //     clicked.current.parent.parent.localToWorld(p.set(0, GOLDENRATIO / 2, 1.25))
  //     clicked.current.parent.parent.getWorldQuaternion(q)
  //   } else {
  //     p.set(0, 0, 162)
  //     q.identity()
  //   }
  // })

  useFrame((stateFrame) => {
    if (clicked.current) {
      stateFrame.camera.position.lerp(
        p.set(stateFrame.camera.position.x, stateFrame.camera.position.y, clicked.current.position.z + 0.5),
        0.025
      )
    }
  })

  // useCursor(hovered)
  // useFrame((s) => {
  //   vec.set(s.size.width / 2 + (s.mouse.x * s.size.width) / 2, s.size.height / 2 - (s.mouse.y * s.size.height) / 2)
  //   if (Math.abs(pos.x - vec.x) > eps || Math.abs(pos.y - vec.y) > eps) {
  //     pos.lerp(vec, speed)
  //     outer.current.style.transform = `translate3d(${pos.x - 55}px,${pos.y - 55}px,0)`
  //   }
  // })

  return (
    // onPointerOver={(e) => (e.stopPropagation(), hover(e.object))} onPointerOut={(e) => hover(null)}
    <group
      ref={ref}
      onWheel={(e) => {
        clicked.current = null
        sections.map(({ section, objName }) => {
          if (objName) {
            e.camera.updateMatrixWorld() // make sure the camera matrix is updated
            e.camera.matrixWorldInverse.copy(e.camera.matrixWorld).invert()
            cameraViewProjectionMatrix.multiplyMatrices(e.camera.projectionMatrix, e.camera.matrixWorldInverse)
            frustum.setFromProjectionMatrix(cameraViewProjectionMatrix)
            bbox.setFromObject(ref.current.getObjectByName(objName))
            if (frustum.intersectsBox(bbox)) {
              setCurrentSection(section)
            }
          }
          return false
        })
      }}>
      {children}
      {/* Everything we'll put into the tunnels "In" will be projected
          outwards to the tunnels "Out". That means we can write dom nodes
          from within r3f, with full access to canvas state! */}
      <dom.In>
        <div ref={outer} className='pointer-events-none absolute left-0 top-0'>
          <div
            className='w-[120px] h-[120px] border-2 border-solid border-[orange] rounded-[50%]'
            style={{
              transition: `all 350ms cubic-bezier(0, 0.28, 0, 0.77)`,
              transform: `scale(${hovered ? 0.55 : 0.29})`,
              borderWidth: hovered ? '5px' : '10px',
              borderColor: color,
              backgroundColor: hovered ? 'rgb(51 65 85)' : 'transparent'
            }}
          />
          {/* <div class="cursor-text" style={{ background: color, opacity: hovered ? 1 : 0 }}>
            {color}
          </div> */}
        </div>
        <div className='text-[red] absolute right-[1%] bottom-[1%] flex'>
          {icons.map(({ icon, description, fontSize, link }) => (
            <div
              className='relative flex flex-col items-center group ml-3'
              onMouseEnter={() => hover(true)}
              onMouseLeave={() => hover(false)}>
              <div className='hover:stroke-white stroke-zinc-600'>
                <UseAnimations
                  animation={icon}
                  size={fontSize}
                  // strokeColor={hovered ? '#FFF' : '#49505799'}
                  pathCss='stroke: inherit;'
                  wrapperStyle={{ stroke: 'inherit' }}
                  onClick={() => openInNewTab(link)}
                />
              </div>

              <div className='absolute bottom-1 flex flex-col items-center hidden mb-10 group-hover:flex'>
                <span className='relative z-10 p-2 text-xs leading-none text-white whitespace-nowrap bg-gray-800 shadow-lg'>
                  {description}
                </span>
                <div className='w-3 h-3 -mt-2 rotate-45 bg-gray-800' />
              </div>
            </div>
          ))}
        </div>
        <div className='text-[red] absolute right-[1%] top-[2%] flex '>
          {sections.map(({ section, objName }) => (
            <div
              role='presentation'
              className='relative flex flex-col items-center mx-3'
              onMouseEnter={() => hover(true)}
              onMouseLeave={() => hover(false)}
              onClick={() => {
                clickSound.play()
                clicked.current = ref.current.getObjectByName(objName)
                setCurrentSection(section)
                // api.refresh(clicked.current).fit()
              }}>
              <div
                className={`font-mono text-base hover:text-white ${
                  currentSection === section ? 'text-red-500' : 'text-zinc-500'
                } hover:underline underline-offset-4 decoration-3`}>
                {section}
              </div>
            </div>
          ))}
        </div>
      </dom.In>
    </group>
  )
}

export default Cursor
