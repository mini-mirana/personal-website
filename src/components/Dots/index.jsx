import React, { useMemo, useRef } from 'react'
import { useFrame } from '@react-three/fiber'
import * as THREE from 'three'

// NON-INTERACTIVE VERSION

// Equation from https://dsp.stackexchange.com/a/56529
// Visualized here https://www.desmos.com/calculator/uakymahh4u
const roundedSquareWave = (t, delta, a, f) => {
  return ((2 * a) / Math.PI) * Math.atan(Math.sin(2 * Math.PI * t * f) / delta)
}

export function Dots() {
  const ref = useRef() // Reference to our InstancedMesh
  const { vec, transform, positions, distances } = useMemo(() => {
    // Variables for intermediary calculations
    const v = new THREE.Vector3()
    const t = new THREE.Matrix4()

    // Precompute randomized initial positions (array of Vector3)
    const pos = [...Array(10000)].map((_, i) => {
      const p = new THREE.Vector3()
      // Place in a grid
      p.x = (i % 100) - 50
      p.y = Math.floor(i / 100) - 50
      // Place in a grid
      p.x /= 2
      p.y /= 2

      // Offset every other column (hexagonal pattern)
      // p.y += (i % 2) * 0.5

      // Add some noise
      // p.x += Math.random() * 0.3
      // p.y += Math.random() * 0.3
      return p
    })

    // Precompute initial distances with octagonal offset
    const right = new THREE.Vector3(1, 0, 0)
    const d = pos.map((i) => {
      // See https://www.desmos.com/calculator/4eiqbvjdzm
      return i.length() + Math.cos(i.angleTo(right) * 8) * 0.5
    })
    return { v, t, pos, d }
  }, [])
  useFrame(({ clock }) => {
    for (let i = 0; i < 10000; i += 1) {
      const dist = distances[i]

      // Distance affects the wave phase
      const t = clock.elapsedTime - dist / 25

      // Oscillates between -0.4 and +0.4 with period of 3.8 seconds
      const wave = roundedSquareWave(t, 0.15 + (0.2 * dist) / 72, 0.4, 1 / 3.8)

      // Scale initial position by our oscillator
      vec.copy(positions[i]).multiplyScalar(wave + 1.3)

      // Apply the Vector3 position to the Matrix4
      transform.setPosition(vec)

      // Update Matrix4 for this instance
      ref.current.setMatrixAt(i, transform)
    }
    ref.current.instanceMatrix.needsUpdate = true
  })
  return (
    <instancedMesh ref={ref} args={[null, null, 20000]}>
      <sphereBufferGeometry args={[0.01]} /> {/* Circle with radius of 0.15 */}
      <meshBasicMaterial /> {/* Default to white color */}
    </instancedMesh>
  )
}

export default Dots
