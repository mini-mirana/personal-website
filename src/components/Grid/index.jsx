import { useThree } from '@react-three/fiber'
import { useAspect } from '@react-three/drei'
import { Flex, Box } from '@react-three/flex'
import { animated, useSprings } from '@react-spring/three'
import { Text } from '../Text'
import { TextMesh } from '../TextMesh'
import defaultFont from '../../assets/NoirPro-Bold.json' /* three/examples/fonts/helvetiker_bold.typeface.json */

export function Grid({ startZ = undefined, distance = 8, reverse = false, startReverse = false, content = [] }) {
  const { size } = useThree()
  const [vpWidth, vpHeight] = useAspect(size.width, size.height)

  const springs = []
  const apis = []
  content.forEach((c) => {
    const us = useSprings(c.config.rowNumber * c.config.columnNumber, () => ({
      scale: [1, 1, 1],
      opacity: 0.2,
      loop: true,
      config: { mass: 5, tension: 350, friction: 40 }
    }))
    springs.push(us[0])
    apis.push(us[1])
  })

  // useFrame(({ clock }) => {
  //   // sphereRef.current.rotation.y += -0.01;
  //   matRef.current.emissiveIntensity = Math.abs(Math.sin(clock.elapsedTime * 0.5))
  // })

  return (
    <>
      {content.map((c, i) => (
        <group
          position={[0, 0, startZ - i * distance]}
          rotation={(reverse && ((startReverse && [0, 0, (i + 1) * Math.PI]) || [0, 0, i * Math.PI])) || [0, 0, 0]}>
          <Flex
            key={`${c.title.content}${c.title?.font}`}
            size={[vpWidth, vpHeight, 0]}
            position={[-vpWidth / 2, vpHeight / 2, 0]}
            alignItems='center'
            justifyContent='center'>
            <Text
              font={
                c.title?.font ||
                'https://fonts.gstatic.com/s/raleway/v17/1Ptxg8zYS_SKggPN4iEgvnHyvveLxVvao7CIPrcVIT9d0c8.woff'
              }
              fontSize={c.title?.fontSize || 0.12}
              letterSpacing={0.1}
              textAlign='center'
              anchorX='center'
              position-x={vpWidth / 2}
              anchorY='center'
              position-y={
                -vpHeight / 2 +
                (c.config.rowNumber * (c.config?.boxHeight || 0.5) +
                  c.config.rowNumber * (2 * (c.config?.boxMargin || 0.05))) /
                  2 +
                parseFloat(c.title?.fontSize || 0.12) +
                0.06
              }>
              {c.title.content}
              <meshStandardMaterial color={c.title?.color || '#fff'} />
            </Text>
            <Box>
              <Box
                flexDirection='row'
                flexWrap='wrap'
                width={
                  c.config.columnNumber * (c.config?.boxWidth || 0.5) +
                  c.config.columnNumber * (2 * (c.config?.boxMargin || 0.05))
                }
                flexGrow={1}>
                {springs[i].map(({ scale, opacity }, j) => (
                  <group
                    onPointerOver={() =>
                      apis[i].start((index) => {
                        if (index === j) {
                          return { scale: [1.1, 1.1, 1], opacity: 0.5 }
                        }
                        return { scale: [1, 1, 1], opacity: 0.2 }
                      })
                    }
                    onClick={() => {
                      apis[i].start((index) => {
                        if (index === j) {
                          if (opacity.animation.to === 0.2) {
                            return { scale: [1.1, 1.1, 1], opacity: 0.5 }
                          }
                        }
                        return { scale: [1, 1, 1], opacity: 0.2 }
                      })
                    }}
                    onPointerLeave={() => apis[i].start(() => ({ scale: [1, 1, 1], opacity: 0.2 }))}>
                    <Box width={c.config?.boxWidth || 0.5} margin={c.config?.boxMargin || 0.05}>
                      <animated.mesh
                        scale={scale}
                        position={[(c.config?.boxWidth || 0.5) / 2, -(c.config?.boxHeight || 0.5) / 2, 0]}>
                        <planeBufferGeometry args={[c.config?.boxWidth || 0.5, c.config?.boxHeight || 0.5]} />
                        <animated.meshPhysicalMaterial
                          color='red'
                          roughness={0.5}
                          metalness={1}
                          emissive='blue'
                          emissiveIntensity={0.5}
                          transparent
                          opacity={opacity}
                        />
                        {/* <meshStandardMaterial args={[{ transparent: true, opacity: 0.1 }]} /> */}
                        {/* <shaderMaterial ref={material} attach='material' {...data} /> */}
                      </animated.mesh>
                      {c.text[j].type === 'text' && (
                        <Text
                          font={
                            c.text[j]?.font ||
                            'https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.9.0/webfonts/fa-brands-400.woff'
                          }
                          fontSize={c.text[j]?.fontSize || 0.14}
                          letterSpacing={0.1}
                          textAlign='center'
                          anchorX='center'
                          position-x={(c.config?.boxWidth || 0.5) / 2}
                          anchorY='middle'
                          position-y={-(c.config?.boxHeight || 0.5) / 2}>
                          {c.text[j].content}
                          <meshStandardMaterial color={c.text[j]?.color || '#fff'} />
                        </Text>
                      )}
                      {c.text[j].type === 'textMesh' && (
                        <TextMesh
                          position-x={(c.config?.boxWidth || 0.5) / 2}
                          position-y={-(c.config?.boxHeight || 0.5) / 2}
                          hAlign='right'
                          fontUrl={c.text[j]?.font || defaultFont}
                          fontConfig={{
                            size: c.text[j]?.size || 7,
                            height: c.text[j]?.height || 0.08,
                            curveSegments: c.text[j]?.curveSegments || 32,
                            bevelEnabled: true,
                            bevelThickness: c.text[j]?.bevelThickness || 0.1,
                            bevelSize: c.text[j]?.bevelSize || 0.1,
                            bevelOffset: c.text[j]?.bevelOffset || 0,
                            bevelSegments: c.text[j]?.bevelSegments || 2
                          }}>
                          {c.text[j].content}
                        </TextMesh>
                      )}
                    </Box>
                  </group>
                ))}
              </Box>
            </Box>
            {/* ))} */}
          </Flex>
        </group>
      ))}
    </>
  )
}

export default Grid
