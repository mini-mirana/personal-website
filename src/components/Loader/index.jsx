import * as React from 'react'
import { useProgress } from '@react-three/drei'
import Lottie from 'react-lottie'
import { ArwesThemeProvider, StylesBaseline, Text as ArwesText } from '@arwes/core'
import { BleepsProvider } from '@arwes/sounds'
import { AnimatorGeneralProvider, Animator } from '@arwes/animation'
import animationData from '../../../public/lotties/vortex-loader.json'

const defaultOptions = {
  loop: true,
  autoplay: true,
  animationData,
  rendererSettings: {
    preserveAspectRatio: 'xMidYMid slice'
  }
}

const defaultDataInterpolation = (p) => `Loading ${p.toFixed(2)}%`

export function Loader({
  // containerStyles,
  // innerStyles,
  // barStyles,
  // dataStyles,
  dataInterpolation = defaultDataInterpolation
  // initialState = (active) => active,
}) {
  const { active, progress } = useProgress()
  const progressRef = React.useRef(0)
  const rafRef = React.useRef(0)
  const progressSpanRef = React.useRef(null)
  const [shown, setShown] = React.useState(true)
  const [clicked, setClicked] = React.useState(false)
  const [keepActive, setKeepActive] = React.useState(false)

  React.useEffect(() => {
    let t
    if (clicked === shown)
      t = setTimeout(() => {
        new Audio('/sound/sweetchoff.wav').play()
        setShown(!clicked)
      }, 2000)
    return () => clearTimeout(t)
  }, [shown, clicked])

  React.useEffect(() => {
    setKeepActive(true)
  }, [active])

  const updateProgress = React.useCallback(() => {
    if (!progressSpanRef.current) return
    progressRef.current += (progress - progressRef.current) / 2
    if (progressRef.current > 0.95 * progress || progress === 100) progressRef.current = progress
    progressSpanRef.current.innerText = dataInterpolation(progressRef.current)
    if (progressRef.current < progress) rafRef.current = requestAnimationFrame(updateProgress)
  }, [dataInterpolation, progress])

  React.useEffect(() => {
    updateProgress()
    return () => cancelAnimationFrame(rafRef.current)
  }, [updateProgress])

  return shown && process.env.NEXT_PUBLIC_APP_ENV === undefined ? (
    <ArwesThemeProvider>
      <StylesBaseline />
      <BleepsProvider
        audioSettings={{
          common: { volume: 0.25 }
        }}
        playersSettings={{
          object: { src: ['/sound/object.mp3'] },
          assemble: { src: ['/sound/assemble.mp3'], loop: true },
          type: { src: ['/sound/type.mp3'], loop: true },
          click: { src: ['/sound/click.mp3'] }
        }}
        bleepsSettings={{
          object: { player: 'object' },
          assemble: { player: 'assemble' },
          type: { player: 'type' },
          click: { player: 'click' }
        }}>
        <div
          className='w-screen h-screen bg-black absolute top-0 left-0 z-[1000]'
          role='presentation'
          onClick={() => {
            if (keepActive) {
              setClicked(true)
            }
          }}>
          <div className='w-full h-full flex flex-col	 justify-center items-center'>
            <div className='w-[40%] md:w-[20%]'>
              <Lottie options={defaultOptions} />
            </div>
            {keepActive && (
              <AnimatorGeneralProvider animator={{ duration: { enter: 2500, exit: 100 } }}>
                <Animator animator={{ animation: true, manager: 'stagger' }}>
                  <ArwesText className='ml-[1%]'>Tap anywhere to start</ArwesText>
                </Animator>
              </AnimatorGeneralProvider>
            )}
          </div>
        </div>
      </BleepsProvider>
    </ArwesThemeProvider>
  ) : // <div style={{ ...styles.container, opacity: active ? 1 : 0, ...containerStyles }}>
  //   <div>
  //     <div style={{ ...styles.inner, ...innerStyles }}>
  //       <div style={{ ...styles.bar, transform: `scaleX(${progress / 100})`, ...barStyles }}></div>
  //       <span ref={progressSpanRef} style={{ ...styles.data, ...dataStyles }} />
  //     </div>
  //   </div>
  // </div>
  null
}

// const styles = {
//   container: {
//     position: 'absolute',
//     top: 0,
//     left: 0,
//     width: '100%',
//     height: '100%',
//     background: '#171717',
//     display: 'flex',
//     alignItems: 'center',
//     justifyContent: 'center',
//     transition: 'opacity 300ms ease',
//     zIndex: 1000,
//   },
//   inner: {
//     width: 100,
//     height: 3,
//     background: '#272727',
//     textAlign: 'center',
//   },
//   bar: {
//     height: 3,
//     width: '100%',
//     background: 'white',
//     transition: 'transform 200ms',
//     transformOrigin: 'left center',
//   },
//   data: {
//     display: 'inline-block',
//     position: 'relative',
//     fontVariantNumeric: 'tabular-nums',
//     marginTop: '0.8em',
//     color: '#f0f0f0',
//     fontSize: '0.6em',
//     fontFamily: `-apple-system, BlinkMacSystemFont, "Inter", "Segoe UI", "Helvetica Neue", Helvetica, Arial, Roboto, Ubuntu, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"`,
//     whiteSpace: 'nowrap',
//   },
// }

export default Loader
