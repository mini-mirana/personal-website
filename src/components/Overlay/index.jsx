import { useFrame, useThree, createPortal } from '@react-three/fiber'
import { OrthographicCamera, useCamera } from '@react-three/drei'
import React, { useMemo, useRef } from 'react'
import { Scene } from 'three'

export function Overlay({ children, position }) {
  const { scene } = useThree()
  const virtualScene = useMemo(() => new Scene())
  const virtualCam = useRef()
  const ref = useRef()

  useFrame(({ gl, camera }) => {
    // ref.current.quaternion.setFromEuler(new THREE.Euler(1, 1, 1))
    // eslint-disable-next-line no-param-reassign
    gl.autoClear = true
    gl.render(scene, camera)
    // eslint-disable-next-line no-param-reassign
    gl.autoClear = false
    gl.clearDepth()
    gl.render(virtualScene, virtualCam.current)
  }, 10)

  const textMesh = useRef()
  useFrame(({ clock }) => {
    const r = Math.sin(clock.getElapsedTime())
    textMesh.current.rotation.x = r * 0.2
    textMesh.current.rotation.y = r * 0.2
    textMesh.current.rotation.z = r * 0.2
  })

  return createPortal(
    <>
      <OrthographicCamera ref={virtualCam} makeDefault={false} position={[0, 0, 170]} rotation={[0, 0, 0]} />
      <mesh ref={ref} raycast={useCamera(virtualCam)} position={position}>
        <group ref={textMesh}>{children}</group>
      </mesh>
    </>,
    virtualScene
  )
}

export default Overlay
