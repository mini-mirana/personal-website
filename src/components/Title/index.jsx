import { Flex, Box } from '@react-three/flex'
import { Text } from '../Text'

export function Title({ ...props }) {
  return (
    <group rotation={[0, 0, Math.PI]}>
      <Flex {...props}>
        <Box flexDirection='column' alignItems='center' justifyContent='center' width='100%' height='100%'>
          <Box margin={0.05}>
            <Text fontSize={0.5} letterSpacing={0.1} textAlign='center'>
              REACT
              <meshStandardMaterial />
            </Text>
          </Box>
          <Box margin={0.05}>
            <Text fontSize={0.5} letterSpacing={0.1} textAlign='center'>
              THREE
              <meshStandardMaterial />
            </Text>
          </Box>
          <Box margin={0.05}>
            <Text fontSize={0.5} letterSpacing={0.1} textAlign='center'>
              FIBER
              <meshStandardMaterial />
            </Text>
          </Box>
        </Box>
      </Flex>
    </group>
  )
}

export default Title
