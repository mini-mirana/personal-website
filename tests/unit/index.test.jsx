import React from 'react'
import ReactThreeTestRenderer from '@react-three/test-renderer'
import { Title } from '../../src/components/Title'

// eslint-disable-next-line global-require
jest.mock('scheduler', () => require('scheduler/unstable_mock'))

// const findByType = async (renderer, type) => {
//   return renderer.toTree()[0].children.find(async (mesh) => (await mesh.type) === type)
// }

// const findByText = async (renderer, text) => {
//   return renderer.toTree()[0].children.find(async (mesh) => {
//     console.log(mesh)
//     return (await mesh?.props.text) === text
//   })
// }

describe('Title', () => {
  it('renders Title component', async () => {
    const renderer = await ReactThreeTestRenderer.create(<Title />)

    // assertions using the TestInstance & Scene Graph
    expect(renderer.toTree()).toMatchSnapshot()
  })
})
