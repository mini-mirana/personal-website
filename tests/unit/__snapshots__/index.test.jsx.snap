// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title renders Title component 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "props": Object {
                  "attach": "material",
                },
                "type": "meshStandardMaterial",
              },
            ],
            "props": Object {
              "anchorX": "left",
              "anchorY": "top",
              "font": "https://fonts.gstatic.com/s/raleway/v17/1Ptxg8zYS_SKggPN4iEgvnHyvveLxVvao7CIPrcVIT9d0c8.woff",
              "fontSize": 0.5,
              "letterSpacing": 0.1,
              "maxWidth": undefined,
              "text": "REACT",
              "textAlign": "center",
            },
            "type": "mesh",
          },
        ],
        "props": Object {},
        "type": "group",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "props": Object {
                  "attach": "material",
                },
                "type": "meshStandardMaterial",
              },
            ],
            "props": Object {
              "anchorX": "left",
              "anchorY": "top",
              "font": "https://fonts.gstatic.com/s/raleway/v17/1Ptxg8zYS_SKggPN4iEgvnHyvveLxVvao7CIPrcVIT9d0c8.woff",
              "fontSize": 0.5,
              "letterSpacing": 0.1,
              "maxWidth": undefined,
              "text": "THREE",
              "textAlign": "center",
            },
            "type": "mesh",
          },
        ],
        "props": Object {},
        "type": "group",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "props": Object {
                  "attach": "material",
                },
                "type": "meshStandardMaterial",
              },
            ],
            "props": Object {
              "anchorX": "left",
              "anchorY": "top",
              "font": "https://fonts.gstatic.com/s/raleway/v17/1Ptxg8zYS_SKggPN4iEgvnHyvveLxVvao7CIPrcVIT9d0c8.woff",
              "fontSize": 0.5,
              "letterSpacing": 0.1,
              "maxWidth": undefined,
              "text": "FIBER",
              "textAlign": "center",
            },
            "type": "mesh",
          },
        ],
        "props": Object {},
        "type": "group",
      },
    ],
    "props": Object {},
    "type": "group",
  },
]
`;
